PyDynamic/uncertainty/propagate_DFT.py

```diff
@@ -1,776 +1,908 @@
-			phase values
-		UAP: np.ndarray of shape (M, 3N)
-			diagonals of the covariance matrices: [diag(UPP), diag(UAA), diag(UPA)]
-	"""
-	M, nx = x.shape
-	assert(len(Ux)==M)
-	N = nx//2+1
-	if not isinstance(selector, np.ndarray):
-		selector = np.arange(nx//2+1)
-	ns = len(selector)
-
-	A = np.zeros((M,ns))
-	P = np.zeros_like(A)
-	UAP = np.zeros((M, 3*ns))
-	CxCos = None
-	CxSin = None
-	for m in range(M):
-		F, UF, CX = GUM_DFT(x[m,:], Ux[m], CxCos, CxSin, returnC=True)
-		CxCos = CX["CxCos"]
-		CxSin = CX["CxSin"]
-		A_m, P_m, UAP_m = DFT2AmpPhase(F, UF, keep_sparse=True)
-		A[m,:] = A_m[selector]
-		P[m,:] = P_m[selector]
-		UAP[m,:ns] = UAP_m.data[0][:N][selector]
-		UAP[m,ns:2*ns] = UAP_m.data[1][UAP_m.offsets[1]:2*N+UAP_m.offsets[1]][selector]
-		UAP[m, 2*ns:] = UAP_m.data[0][N:][selector]
-
-	return A, P, UAP
-
-def AmpPhase2Time(A,P,UAP):
-	"""Transformation from amplitude and phase to time domain
-
-	GUM propagation of covariance matrix UAP associated with DFT amplitude A and phase P to the result of
-	the inverse DFT. Uncertainty UAP is assumed to be given for amplitude and phase with blocks:
-	UAP = [[u(A,A), u(A,P)],[u(P,A),u(P,P)]]
-
-	Parameters
-	----------
-		A: np.ndarray of shape (N,)
-			vector of amplitude values
-		P: np.ndarray of shape (N,)
-			vector of phase values (in rad)
-		UAP: np.ndarray of shape (2N,2N)
-			covariance matrix associated with [A,P]
-
-	Returns
-	-------
-		x: np.ndarray
-			vector of time domain values
-		Ux: np.ndarray
-			covariance matrix associated with x
-	"""
-
-	N = UAP.shape[0]-2
-	assert(np.mod(N,2)==0)
-	beta = 2*np.pi*np.arange(N)/N
-
-	# calculate inverse DFT
-	F = A*np.exp(1j*P)
-	x = np.fft.irfft(F)
-
-	Pf = np.r_[P,-P[-2:0:-1]]		# phase values to take into account symmetric part
-	Cc = np.zeros((N,N//2+1))		# sensitivities wrt cosinus part
-	Cc[:,0] = np.cos(P[0]); Cc[:,-1] = np.cos(P[-1]+np.pi*np.arange(N))
-	for k in range(1,N//2):
-		Cc[:,k] = 2*np.cos(Pf[k]+k*beta)
-
-	Cs = np.zeros((N,N//2+1))		# sensitivities wrt sinus part
-	Cs[:,0] = -A[0]*np.sin(P[0]); Cs[:,-1] = -A[-1]*np.sin(P[-1]+np.pi*np.arange(N))
-	for k in range(1,N//2):
-		Cs[:,k] = -A[k]*2*np.sin(Pf[k]+k*beta)
-
-	# calculate blocks of uncertainty matrix
-	if len(UAP.shape)==1:
-		AA = UAP[:N//2+1]
-		PP = UAP[N//2+1:]
-		Ux = np.dot(Cc,prod(AA,Cc.T)) + np.dot(Cs,prod(PP,Cs.T))
-	else:
-		if isinstance(UAP,sparse.dia_matrix):
-			nrows = UAP.shape[0]
-			n = nrows//2
-			offset= UAP.offsets
-			diags = UAP.data
-			AA = diags[0][:n]
-			AP = diags[1][offset[1]:nrows+offset[1]]
-			PP = diags[0][n:]
-			Ux = np.dot(Cc,prod(AA,Cc.T)) + 2*np.dot(Cc,prod(AP,Cs.T)) + np.dot(Cs,prod(PP,Cs.T))
-		else:
-			AA = UAP[:N//2+1,:N//2+1]
-			AP = UAP[:N//2+1, N//2+1:]
-			PP = UAP[N//2+1:,N//2+1:]
-			# propagate uncertainties
-			Ux = np.dot(Cc,np.dot(AA,Cc.T)) + 2*np.dot(Cc,np.dot(AP,Cs.T)) + np.dot(Cs,np.dot(PP,Cs.T))
-
-	return x,Ux/N**2
+    """Transformation from time domain to amplitude and phase
```

Comment:
In many of the other functions, >>"""<< was followed by a newline. Maybe here to?

Reply:
see above